### PR TITLE
drivers: regulator: add regulator shell

### PR DESCRIPTION
**Pull request by danieldegrasse**
Add a regulator shell. This shell supports enabling and disabling regulators, and also supports the extended PMIC regulator API when a PMIC device is present.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

### CODE REVIEW

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Why not add a generic PMIC shell (or better yet, regulator shell) under `drivers/regulator/` that users can enable where needed?

**Comment by danieldegrasse** [comment]
PR updated to move the PMIC shell to a shell available within the regulators driver folder. The shell still only supports PMIC regulators, because there are parts of the PMIC API that are not present on standard regulators. The sample has been changed to simply enable this shell.

**Comment by danieldegrasse** [comment]
@henrikbrixandersen Could you take another look at this?

File: drivers/regulator/Kconfig.pmic [hunk 1]
```diff
@@ -9,12 +9,22 @@ config REGULATOR_PMIC
...
 	default 75
 	help
 	  I2C based regulator init priority.
 	  Must be greater than KERNEL_INIT_PRIORITY_DEVICE so I2C is initialized,
 	  and less than SDMMC_INIT_PRIORITY, since the SDMMC driver uses an I2C
 	  PMIC regulator
+
+config REGULATOR_PMIC_SHELL
+	bool "PMIC regulator shell"
+	select SHELL
```
Comment (henrikbrixandersen) [inline drivers/regulator/Kconfig.pmic:25]: I'd use `depends on SHELL` along with `default y` here.

File: drivers/regulator/pmic_shell.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+#include <zephyr/zephyr.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/drivers/regulator/consumer.h>
+
+LOG_MODULE_REGISTER(regulator_shell, CONFIG_REGULATOR_LOG_LEVEL);
+
+#define GET_REGULATOR(node) DEVICE_DT_GET(node),
```
Comment (henrikbrixandersen) [inline drivers/regulator/pmic_shell.c:15]: What's the point here? You're just making an alias for an established, well-known macro?

File: drivers/regulator/pmic_shell.c [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+static const struct device *supported_pmics[] = {
+	DT_FOREACH_STATUS_OKAY(regulator_pmic, GET_REGULATOR)
+	};
+static const struct device *pmic_dev;
```
Comment (henrikbrixandersen) [inline drivers/regulator/pmic_shell.c:20]: Why not use a `<device>` argument to all `regulator` shell commands as done in other shells? There's no need to make this array at compile-time. Technically, other regulator drivers can than PMIC devices can be controlled via the exact same regulator API calls.

File: drivers/regulator/pmic_shell.c [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+#define MAX_REGULATOR_VOLTAGE 5000000 /* Max voltage of 5V */
```
Comment (henrikbrixandersen) [inline drivers/regulator/pmic_shell.c:22]: Why?

File: drivers/regulator/pmic_shell.c [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+static int cmd_pmic_en(const struct shell *sh, size_t argc, char **argv)
+{
+	struct onoff_client cli;
+
+	if (pmic_dev == NULL) {
+		shell_print(sh, "No PMIC device has been selected");
```
Comment (henrikbrixandersen) [inline drivers/regulator/pmic_shell.c:29]: You need to use `shell_error()` for error printouts.

File: drivers/regulator/pmic_shell.c [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		return -ENODEV;
+	}
+	if (!device_is_ready(pmic_dev)) {
+		shell_print(sh, "PMIC device not ready");
+		return -ENODEV;
+	}
+	shell_print(sh, "Enabling PMIC");
+	if (regulator_enable(pmic_dev, &cli) < 0) {
+		shell_print(sh, "Failed  to enable PMIC");
+		return -EACCES;
```
Comment (henrikbrixandersen) [inline drivers/regulator/pmic_shell.c:39]: Why `-EACCESS`? Why not propagate the error returned (and mention it in the error output for easier debugging)?

File: drivers/regulator/pmic_shell.c [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	return 0;
+}
+
+static int cmd_list_devices(const struct shell *sh, size_t argc, char **argv)
+{
+	for (int i = 0; i < ARRAY_SIZE(supported_pmics); i++) {
+		shell_print(sh, "%s", supported_pmics[i]->name);
+	}
+	return 0;
+}
```
Comment (henrikbrixandersen) [inline drivers/regulator/pmic_shell.c:155]: This is what `device list` is for.

File: drivers/regulator/pmic_shell.c [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+	SHELL_CMD(disable, NULL, "Disable buck regulator", cmd_pmic_dis),
+	SHELL_CMD(set_vol, NULL, "Set voltage (in mV)", cmd_pmic_set_vol),
+	SHELL_CMD(set_current, NULL, "Set current limit( in mA)",
+		cmd_pmic_set_ilim),
+	SHELL_CMD(set_device, NULL, "Set regulator device", cmd_pmic_set_dev),
+	SHELL_CMD(list_devices, NULL, "List enabled regulators",
+		cmd_list_devices),
+	SHELL_SUBCMD_SET_END
+);
+SHELL_CMD_REGISTER(pmic, &pmic_set, "PMIC Management", NULL);
```
Comment (henrikbrixandersen) [inline drivers/regulator/pmic_shell.c:169]: I would prefer a generic top-level `regulator` command with `enable` and `disable` commands - and a conditional `pmic` subcommand for the PMIC specific commands. This will make the shell module useful for non-PMIC regulators as well.

File: samples/drivers/pmic_shell/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(pmic_shell)
```
Comment (henrikbrixandersen) [inline samples/drivers/pmic_shell/CMakeLists.txt:6]: We usually don't add a sample just for a shell module as it can easily be enabled where needed.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/regulator/Kconfig.pmic (hunk 1 above)
Comment (danieldegrasse) [reply drivers/regulator/Kconfig.pmic]: Fixed in latest push

File: drivers/regulator/pmic_shell.c (hunk 2 above)
Comment (danieldegrasse) [reply drivers/regulator/pmic_shell.c]: yeah, this is wrong- I've removed it

File: drivers/regulator/pmic_shell.c (hunk 3 above)
Comment (danieldegrasse) [reply drivers/regulator/pmic_shell.c]: Changed the commands to use a `device` argument directly

File: drivers/regulator/pmic_shell.c (hunk 4 above)
Comment (danieldegrasse) [reply drivers/regulator/pmic_shell.c]: This is legacy, since this shell was initially used to test the PCA9420, a PMIC regulator that won't go above that voltage. I've made the max voltage an argument to the shell command to set the PMIC voltage

File: drivers/regulator/pmic_shell.c (hunk 5 above)
Comment (danieldegrasse) [reply drivers/regulator/pmic_shell.c]: Thanks, I've update this accordingly

File: drivers/regulator/pmic_shell.c (hunk 6 above)
Comment (danieldegrasse) [reply drivers/regulator/pmic_shell.c]: Good point- just tried to choose a logical error here. I've changed this to propagate the error

File: drivers/regulator/pmic_shell.c (hunk 8 above)
Comment (danieldegrasse) [reply drivers/regulator/pmic_shell.c]: I've updated the shell to use this approach, so that generic regulators will work

File: samples/drivers/pmic_shell/CMakeLists.txt (hunk 9 above)
Comment (danieldegrasse) [reply samples/drivers/pmic_shell/CMakeLists.txt]: Removed sample in latest push

**Review by carlescufi — DISMISSED** [review]
LGTM

File: drivers/regulator/Kconfig [hunk 10]
```diff
@@ -8,6 +8,13 @@ menuconfig REGULATOR
 
 if REGULATOR
 
+config REGULATOR_SHELL
+	bool "Regulator shell"
+	depends on SHELL
```
Comment (henrikbrixandersen) [inline drivers/regulator/Kconfig:13]:
```suggestion
	default y
	depends on SHELL
```

File: drivers/regulator/Kconfig.pmic [hunk 11]
```diff
@@ -9,12 +9,15 @@ config REGULATOR_PMIC
...
 	int "Regulator Init priority"
-	depends on REGULATOR_PMIC
 	default 75
 	help
 	  I2C based regulator init priority.
 	  Must be greater than KERNEL_INIT_PRIORITY_DEVICE so I2C is initialized,
 	  and less than SDMMC_INIT_PRIORITY, since the SDMMC driver uses an I2C
 	  PMIC regulator
+
+endif # REGULATOR_PMIC
```
Comment (henrikbrixandersen) [inline drivers/regulator/Kconfig.pmic:23]: `depends on REGULATOR_PMIC` is preferred when only a single `config` is affected.

File: drivers/regulator/regulator_shell.c [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+{
+	int lvol, uvol, ret;
+	const struct device *reg_dev;
+
+	if (argc != 4) {
+		shell_print(sh, "usage: %s "
+			"[pmic dev] [lower limit (mV)] [upper limit (mV)]",
+			argv[0]);
+		return -EINVAL;
+	}
```
Comment (henrikbrixandersen) [inline drivers/regulator/regulator_shell.c:79]: Use `SHELL_CMD_ARG` for automatic argc checks.

File: drivers/regulator/regulator_shell.c [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+{
+	int lcur, ucur, ret;
+	const struct device *reg_dev;
+
+	if (argc != 4) {
+		shell_print(sh, "usage: %s "
+			"[pmic dev] [lower limit (uA)] [upper limit (uA)]",
+			argv[0]);
+		return -EINVAL;
+	}
```
Comment (henrikbrixandersen) [inline drivers/regulator/regulator_shell.c:112]: Same here.

File: drivers/regulator/regulator_shell.c [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+static int cmd_reg_en(const struct shell *sh, size_t argc, char **argv)
+{
+	struct onoff_client cli;
+	const struct device *reg_dev;
+	int ret;
+
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [dev name]", argv[0]);
+		return -EINVAL;
+	}
```
Comment (henrikbrixandersen) [inline drivers/regulator/regulator_shell.c:26]: Same here.

File: drivers/regulator/regulator_shell.c [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+static int cmd_reg_dis(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct device *reg_dev;
+	int ret;
+
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [dev name]", argv[0]);
+		return -EINVAL;
+	}
```
Comment (henrikbrixandersen) [inline drivers/regulator/regulator_shell.c:51]: Same here.

File: drivers/regulator/regulator_shell.c [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+	struct onoff_client cli;
+	const struct device *reg_dev;
+	int ret;
+
+	if (argc != 2) {
+		shell_print(sh, "usage: %s [dev name]", argv[0]);
+		return -EINVAL;
+	}
+	reg_dev = device_get_binding(argv[1]);
+	if (reg_dev == NULL) {
```
Comment (henrikbrixandersen) [inline drivers/regulator/regulator_shell.c:28]: You need to use `device_is_ready()` instead of checking against `NULL` all over.

File: drivers/regulator/regulator_pmic.c [hunk 17]
```diff
@@ -273,15 +273,13 @@ static int disable_regulator(const struct device *dev)
...
 	}
-	dis_val = config->enable_inverted ? config->enable_val : 0;
-	rc = regulator_modify_register(config, config->enable_reg,
-			config->enable_mask, dis_val);
-	if (rc) {
-		/* Error writing configs */
-		return onoff_sync_finalize(&data->srv, key, NULL, rc, true);
-	}
-	return onoff_sync_finalize(&data->srv, key, NULL, rc, true);
+	return onoff_sync_finalize(&data->srv, key, NULL, rc, false);
```
Comment (henrikbrixandersen) [inline drivers/regulator/regulator_pmic.c:282]: This is a bug fix and really ought to go into a separate PR.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/regulator/Kconfig (hunk 10 above)
Comment (danieldegrasse) [reply drivers/regulator/Kconfig:15]: Fixed in latest push

File: drivers/regulator/Kconfig.pmic (hunk 11 above)
Comment (danieldegrasse) [reply drivers/regulator/Kconfig.pmic]: Updated to use `depends on REGULATOR_PMIC`

File: drivers/regulator/regulator_shell.c (hunk 12 above)
Comment (danieldegrasse) [reply drivers/regulator/regulator_shell.c]: I've updated all shell commands in this PR to use `SHELL_CMD_ARG`, thanks for the pointer.

File: drivers/regulator/regulator_pmic.c (hunk 17 above)
Comment (danieldegrasse) [reply drivers/regulator/regulator_pmic.c]: Done, PR number is #51092

**Review by henrikbrixandersen — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]
